File: mama/dotnet/src/nunittest/MamaResourcePoolTest.cs
using NUnit.Framework;
using Wombat;

namespace NUnitTest
{
    public class MamaResourcePoolTest
    {
        private string mTransportName = "sub";
        private string mBridgeName;
        private MamaResourcePool mPool;

        class SubscriptionEventHandler : MamaSubscriptionCallback
        {
            public void onCreate(MamaSubscription subscription) {
            }

            public void onError(MamaSubscription subscription, MamaStatus.mamaStatus status, string subject) {
            }

            public void onMsg(MamaSubscription subscription, MamaMsg msg) {
            }

            public void onQuality(MamaSubscription subscription, mamaQuality quality, string subject) {
            }

            public void onRecapRequest(MamaSubscription subscription) {

            }

            public void onGap(MamaSubscription subscription) {

            }

            public void onDestroy(MamaSubscription subscription) {

            }
        }

        [SetUp]
        public void setUp()
        {
            mBridgeName = MamaCommon.middlewareName;
            Mama.open();
            Mama.setProperty("mama.resource_pool.test.bridges", mBridgeName);
            Mama.setProperty("mama.resource_pool.test.default_transport_sub", mTransportName);
            Mama.setProperty("mama.resource_pool.test.default_source_sub", "SOURCE");
            Mama.setProperty(string.Format("mama.{0}.transport.{1}.param", mBridgeName, mTransportName), "value");
            mPool = new MamaResourcePool("test");
        }

        [TearDown]
        public void tearDown()
        {
            mPool.destroy();
            Mama.close();
        }

        [Test]
        public void testCreateNullParams(){
            Assert.Throws<MamaException>(() => new MamaResourcePool(null));
        }

        [Test]
        public void createTransportFromName() {
            MamaTransport transport = mPool.createTransportFromName(mTransportName);
            Assert.NotNull(transport);
            Assert.AreEqual(mTransportName, transport.getName());
        }

        [Test]
        public void createTransportInvalidParameters() {
            Assert.Throws<MamaException>(() => mPool.createTransportFromName(null));
        }

        [Test]
        public void createSubscriptionFromUri() {
            MamaSubscription subscription = mPool.createSubscriptionFromUri (
                    "qpid://sub/SOURCE/test.topic", new SubscriptionEventHandler(), null);
            Assert.AreEqual (subscription.subscSource, "SOURCE");
            Assert.AreEqual (subscription.subscSymbol, "test.topic");
        }

        [Test]
        public void createSubscriptionFromUriInvalidUriPlainString() {
            Assert.Throws<System.UriFormatException>(() => mPool.createSubscriptionFromUri("plainstring", new SubscriptionEventHandler(), null));
        }

        [Test]
        public void createSubscriptionFromUriInvalidUriNullArg() {
            Assert.Throws<MamaException>(() => mPool.createSubscriptionFromUri(null, new SubscriptionEventHandler(), null));
        }

        [Test]
        public void createSubscriptionFromUriInvalidHandler() {
            Assert.Throws<MamaException>(() => mPool.createSubscriptionFromUri("qpid://sub/SOURCE/test.topic", null, null));
        }

        [Test]
        public void createSubscriptionFromComponents() {
            MamaSubscription subscription = mPool.createSubscriptionFromComponents (
                    "sub", "SOURCE", "test.topic", new SubscriptionEventHandler(), null);
            Assert.AreEqual (subscription.subscSource, "SOURCE");
            Assert.AreEqual (subscription.subscSymbol, "test.topic");
        }

        [Test]
        public void createSubscriptionFromTopicWithSource() {
            MamaSubscription subscription = mPool.createSubscriptionFromTopicWithSource (
                    "SOURCE", "test.topic", new SubscriptionEventHandler(), null);
            Assert.AreEqual (subscription.subscSource, "SOURCE");
            Assert.AreEqual (subscription.subscSymbol, "test.topic");
        }

        [Test]
        public void createSubscriptionFromTopic() {
            MamaSubscription subscription = mPool.createSubscriptionFromTopic (
                    "test.topic", new SubscriptionEventHandler(), null);
            Assert.AreEqual (subscription.subscSource, "SOURCE");
            Assert.AreEqual (subscription.subscSymbol, "test.topic");
        }
    }
}

File: mama/c_cpp/src/gunittest/cpp/MamaResourcePoolTest.cpp
#include <gtest/gtest.h>
#include "mama/resourcepool.h"
#include "mama/mama.h"
#include "wombat/property.h"
#include "mama/MamaResourcePool.h"
#include "mama/MamaStatus.h"
#include "mama/mamacpp.h"
#include <cstring>
#include <cstdlib>
#include <memory>

using namespace Wombat;

class MamaResourcePoolTestCPP : public ::testing::Test
{	
protected:
    MamaResourcePoolTestCPP () {};
    ~MamaResourcePoolTestCPP () {};

    void SetUp() override {
        mTransportName = strdup("sub");
        mBridgeName = strdup("qpid");
        Mama::setProperty("mama.resource_pool.test.bridges", mBridgeName);
        Mama::setProperty("mama.resource_pool.test.default_transport_sub", mTransportName);
        Mama::setProperty("mama.resource_pool.test.default_source_sub", "SOURCE");
        char prop[PROPERTY_NAME_MAX_LENGTH];
        snprintf(prop, sizeof(prop), "mama.%s.transport.%s.param", mBridgeName, mTransportName);
        Mama::setProperty(prop, "value");
    };
    void TearDown () override {
        free ((void*)mBridgeName);
        free ((void*)mTransportName);
    };

    const char* mTransportName;
    const char* mBridgeName;
};

TEST_F (MamaResourcePoolTestCPP, CreateDestroy)
{
    auto pool = new MamaResourcePool("test");
    delete pool;
}

TEST_F (MamaResourcePoolTestCPP, CreateNullParams)
{
    ASSERT_THROW (new MamaResourcePool(nullptr), MamaStatus);
}

TEST_F (MamaResourcePoolTestCPP, CreateTransportFromNameDestroyTransport)
{
    std::unique_ptr<MamaResourcePool> pool(new MamaResourcePool("test"));
    auto* transport = pool->createTransportFromName (mTransportName);
    ASSERT_NE (transport, nullptr);
}


TEST_F (MamaResourcePoolTestCPP, CreateTransportInvalidParams)
{
    std::unique_ptr<MamaResourcePool> pool(new MamaResourcePool("test"));
    ASSERT_THROW (pool->createTransportFromName (nullptr), MamaStatus);
}

TEST_F (MamaResourcePoolTestCPP, CreateSubscriptionFromUri)
{
    std::unique_ptr<MamaResourcePool> pool(new MamaResourcePool("test"));
    class SubscriptionEventHandler : public MamaSubscriptionCallback
    {
        void onMsg (MamaSubscription* subscription, MamaMsg& msg) override {}
    };
    SubscriptionEventHandler eventHandler;
    auto *subscription = pool->createSubscriptionFromUri (
        "qpid://sub/SOURCE/test.topic", &eventHandler);
    ASSERT_STREQ (subscription->getSource()->getPublisherSourceName(), "SOURCE");
    ASSERT_STREQ (subscription->getSymbol(), "test.topic");
    ASSERT_STREQ (subscription->getTransport()->getName(), "sub");
}

TEST_F (MamaResourcePoolTestCPP, CreateSubscriptionFromUriInvalidUri)
{
    std::unique_ptr<MamaResourcePool> pool(new MamaResourcePool("test"));
    class SubscriptionEventHandler : public MamaSubscriptionCallback
    {
        void onMsg (MamaSubscription* subscription, MamaMsg& msg) override {}
    };
    SubscriptionEventHandler eventHandler;
    ASSERT_THROW (pool->createSubscriptionFromUri("plainString", &eventHandler), MamaStatus);
    ASSERT_THROW (pool->createSubscriptionFromUri(nullptr, &eventHandler), MamaStatus);
}

TEST_F (MamaResourcePoolTestCPP, CreateSubscriptionFromUriNullArgs) {
    std::unique_ptr<MamaResourcePool> pool(new MamaResourcePool("test"));
    class SubscriptionEventHandler : public MamaSubscriptionCallback
    {
        void onMsg (MamaSubscription* subscription, MamaMsg& msg) override {}
    };
    SubscriptionEventHandler eventHandler;
    ASSERT_THROW (pool->createSubscriptionFromUri("qpid://sub/SOURCE/test.topic", nullptr), MamaStatus);
}

TEST_F (MamaResourcePoolTestCPP, CreateSubscriptionFromComponents)
{
    std::unique_ptr<MamaResourcePool> pool(new MamaResourcePool("test"));
    class SubscriptionEventHandler : public MamaSubscriptionCallback
    {
        void onMsg (MamaSubscription* subscription, MamaMsg& msg) override {}
    };
    SubscriptionEventHandler eventHandler;
    auto *subscription = pool->createSubscriptionFromComponents ("sub", "SOURCE", "test.topic", &eventHandler);
    ASSERT_STREQ (subscription->getSource()->getPublisherSourceName(), "SOURCE");
    ASSERT_STREQ (subscription->getSymbol(), "test.topic");
    ASSERT_STREQ (subscription->getTransport()->getName(), "sub");
}

TEST_F (MamaResourcePoolTestCPP, CreateSubscriptionFromTopicWithSource)
{
    std::unique_ptr<MamaResourcePool> pool(new MamaResourcePool("test"));
    class SubscriptionEventHandler : public MamaSubscriptionCallback
    {
        void onMsg (MamaSubscription* subscription, MamaMsg& msg) override {}
    };
    SubscriptionEventHandler eventHandler;
    auto *subscription = pool->createSubscriptionFromTopicWithSource("SOURCE", "test.topic", &eventHandler);
    ASSERT_STREQ (subscription->getSource()->getPublisherSourceName(), "SOURCE");
    ASSERT_STREQ (subscription->getSymbol(), "test.topic");
    ASSERT_STREQ (subscription->getTransport()->getName(), "sub");
}

TEST_F (MamaResourcePoolTestCPP, CreateSubscriptionFromTopic)
{
    std::unique_ptr<MamaResourcePool> pool(new MamaResourcePool("test"));
    class SubscriptionEventHandler : public MamaSubscriptionCallback
    {
        void onMsg (MamaSubscription* subscription, MamaMsg& msg) override {}
    };
    SubscriptionEventHandler eventHandler;
    auto *subscription = pool->createSubscriptionFromTopic("test.topic", &eventHandler);
    ASSERT_STREQ (subscription->getSource()->getPublisherSourceName(), "SOURCE");
    ASSERT_STREQ (subscription->getSymbol(), "test.topic");
    ASSERT_STREQ (subscription->getTransport()->getName(), "sub");
}
